File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_401_500/Sol496.java
package com.goinhn.thorwcode.leetcode.question._401_500;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

/**
 * 496.下一个更大元素I
 *
 * @author goinhn
 * @date 2022/8/22
 */
public class Sol496 {

    /**
     * 遍历
     * time:O(n)
     * space:O(1)
     */
    private static class Solution1 {

        public int[] nextGreaterElement(int[] nums1, int[] nums2) {
            int[] res = new int[nums1.length];
            for (int i = 0; i < nums1.length; i++) {
                boolean isNum = false;
                boolean isMax = false;
                for (int num2 : nums2) {
                    if (num2 == nums1[i]) {
                        isNum = true;
                    }

                    if (isNum && num2 > nums1[i]) {
                        res[i] = num2;
                        isMax = true;
                        break;
                    }
                }

                if (!isMax) {
                    res[i] = -1;
                }
            }

            return res;
        }

    }

    /**
     * 单调栈
     * time:O(n)
     * space:O(1)
     */
    private static class Solution2 {

        public int[] nextGreaterElement(int[] nums1, int[] nums2) {
            Map<Integer, Integer> map = new HashMap<>();
            Stack<Integer> stack = new Stack<>();
            for (int i = nums2.length - 1; i > -1; i--) {
                while (!stack.isEmpty() && nums2[i] > stack.peek()) {
                    stack.pop();
                }
                map.put(nums2[i], stack.isEmpty() ? -1 : stack.peek());
                stack.push(nums2[i]);
            }

            int[] res = new int[nums1.length];
            for (int i = 0; i < nums1.length; i++) {
                res[i] = map.get(nums1[i]);
            }

            return res;
        }

    }

}

File: leetcode_go/question/_401_500/_496/sol_496.go
// Package _496 下一个更大元素I
// @author goinhn
// @date 2022/8/22
package _496

// 遍历
// time:O(n)
// space:O(1)
func nextGreaterElement1(nums1 []int, nums2 []int) []int {
	res := make([]int, len(nums1))
	for i, num1 := range nums1 {
		isNum := false
		isMax := false
		for _, num2 := range nums2 {
			if num1 == num2 {
				isNum = true
			}

			if isNum && num2 > num1 {
				res[i] = num2
				isMax = true
				break
			}
		}

		if !isMax {
			res[i] = -1
		}
	}

	return res
}

// 单调栈
// time:O(n)
// space:O(1)
func nextGreaterElement2(nums1 []int, nums2 []int) []int {
	hashTable := map[int]int{}
	var stack []int
	for i := len(nums2) - 1; i > -1; i-- {
		for len(stack) > 0 && nums2[i] > stack[len(stack)-1] {
			stack = stack[:len(stack)-1]
		}

		if len(stack) == 0 {
			hashTable[nums2[i]] = -1
		} else {
			hashTable[nums2[i]] = stack[len(stack)-1]
		}

		stack = append(stack, nums2[i])
	}

	res := make([]int, len(nums1))
	for i, num := range nums1 {
		res[i] = hashTable[num]
	}

	return res
}
